### Source code and Supplemental data related to the GLiMMPS manuscript:

-   Keyan Zhao1,2, Zhi-xiang Lu1,2, Juw Won Park1,2, Qing Zhou3
    and Yi Xing1,2,4 \
    GLiMMPS: robust statistical model for regulatory variation of alternative splicing using RNA-seq data\
    Genome Biol. 2013 Jul 22;14(7):R74. doi: 10.1186/gb-2013-14-7-r74.
\
 1 Department of Internal Medicine, University of Iowa, IA 52242, USA.\
 2 Department of Microbiology, Immunology and Molecular Genetics,
University of California, Los Angeles, CA 90095, USA.\
 3 Department of Statistics, University of California, Los Angeles, CA
90095, USA.\
 4 Corresponding author:

\
 Yi Xing\
 Department of Microbiology, Immunology and Molecular Genetics\
 University of California, Los Angeles\
 Los Angeles, CA 90095-7278\
 Phone: 310-825-6806\
 Fax: 310-206-3663\
 Email: [yxing@ucla.edu](mailto:yxing@ucla.edu)

* * * * *

**Source code, turorial and example dataset for GLiMMPS package**

-   [**Source code of the GLiMMPS package**](GLiMMPScode.tar.gz)
-   [**Tutorial on how to use the GLiMMPS package
    (PDF)**](Tutorial_for_GLiMMPS.pdf)
-   [**Example dataset on how to use the GLiMMPS package( file size \~80
    MB)**](https://xinglab.cass.idre.ucla.edu/public/glimmps/example.tar.gz)

* * * * *

**Supplemental dataset for the manuscript**

-   [**Genotype, Junction read counts, and P-values of various models
    from CEU dataset (\~230MB)**](https://xinglab.cass.idre.ucla.edu/public/glimmps/CheungCEU.tar.gz)
-   [**Genotype, Junction read counts, and P-values of various models
    from CEU2 dataset (\~230MB)**](https://xinglab.cass.idre.ucla.edu/public/glimmps/MontgomeryCEU.tar.gz)

* * * * *

(updated on March 12, 2013.)
